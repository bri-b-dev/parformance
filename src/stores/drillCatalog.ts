import { defineStore } from 'pinia'
import type { Drill } from '@/types'

export const useDrillCatalogStore = defineStore('drillCatalog', {
  state: () => ({
    drills: [] as Drill[],
    loaded: false,
    error: null as string | null,
  }),
  actions: {
    async load() {
      if (this.loaded && this.drills.length > 0) return
      try {
        // Prefer dynamic import so Vite bundles JSON for production.
        // In the Vitest environment require() the JSON synchronously to avoid async loader timing issues in tests.
        let mod: any
        if (typeof process !== 'undefined' && (process.env.VITEST || process.env.NODE_ENV === 'test')) {
          // eslint-disable-next-line @typescript-eslint/no-var-requires
          mod = require('../data/drills.json')
        } else {
          // @ts-ignore - dynamic import of JSON handled by Vite during build
          mod = await import('../data/drills.json')
        }
        const data = (mod?.default ?? mod) as Drill[]
        if (!Array.isArray(data) || data.length === 0) {
          throw new Error('No drills found in drills.json')
        }
        this.drills = data
        this.error = null
      } catch (e: any) {
        const message = 'Konnte Drills nicht laden. Bitte sp√§ter erneut versuchen.'
        this.error = e?.message || message
        // Friendly toast event for the UI (browser only)
        if (typeof window !== 'undefined' && typeof window.dispatchEvent === 'function') {
          let evt: any
          try {
            // @ts-ignore CustomEvent may be missing in some environments
            evt = new CustomEvent('toast', { detail: { type: 'error', message } })
          } catch {
            evt = { type: 'toast', detail: { type: 'error', message } }
          }
          window.dispatchEvent(evt as any)
        }
      } finally {
        this.loaded = true
      }
    },
  },
})
