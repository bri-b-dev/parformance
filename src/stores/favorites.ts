import { defineStore } from 'pinia'
import { createPersist } from '@/utils/persistAdapter'

const persist = createPersist('parformance', 1)

export const useFavoritesStore = defineStore('favorites', {
  state: () => ({
    favorites: [] as string[],
    loaded: false,
  }),
  getters: {
    isFavorite: (state) => (drillId: string) => state.favorites.includes(drillId),
    list: (state) => state.favorites,
  },
  actions: {
    async load() {
      const stored = persist.get<string[]>('favorites')
      this.favorites = Array.isArray(stored) ? stored : []
      this.loaded = true
    },
    async persist() {
      persist.setDebounced('favorites', this.favorites)
    },
    async add(drillId: string) {
      if (!this.favorites.includes(drillId)) {
        this.favorites.push(drillId)
        await this.persist()
      }
    },
    async remove(drillId: string) {
      const next = this.favorites.filter(id => id !== drillId)
      if (next.length !== this.favorites.length) {
        this.favorites = next
        await this.persist()
      }
    },
    async toggle(drillId: string) {
      if (this.favorites.includes(drillId)) {
        await this.remove(drillId)
      } else {
        await this.add(drillId)
      }
    },
    async clearAll() {
      this.favorites = []
      await this.persist()
    }
  }
})
