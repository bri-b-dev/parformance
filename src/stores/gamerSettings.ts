import { defineStore } from 'pinia'
import { createPersist } from '@/utils/persistAdapter'

const persist = createPersist('parformance', 1)

export type GamerSettings = {
  mode?: string
  handicapOffset?: number
  collapsed?: boolean
}

export const useGamerSettingsStore = defineStore('gamerSettings', {
  state: () => ({
    // map drillId -> settings
    map: {} as Record<string, GamerSettings>,
    loaded: false,
  }),
  getters: {
    forDrill: (state) => (drillId: string) => state.map[drillId] ?? { collapsed: true, mode: undefined, handicapOffset: 0 },
  },
  actions: {
    async load() {
      const stored = persist.get<Record<string, GamerSettings>>('gamerSettings')
      this.map = stored && typeof stored === 'object' ? stored : {}
      this.loaded = true
    },
    async persist() {
      persist.setDebounced('gamerSettings', this.map)
    },
    async setFor(drillId: string, next: GamerSettings) {
      this.map = { ...this.map, [drillId]: { ...(this.map[drillId] ?? {}), ...next } }
      await this.persist()
    },
    async setCollapsed(drillId: string, collapsed: boolean) {
      await this.setFor(drillId, { collapsed })
    },
    async setMode(drillId: string, mode?: string) {
      await this.setFor(drillId, { mode })
    },
    async setHandicapOffset(drillId: string, handicapOffset?: number) {
      await this.setFor(drillId, { handicapOffset })
    }
  }
})
