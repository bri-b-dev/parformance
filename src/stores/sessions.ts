import { defineStore } from 'pinia'
import type { Session } from '@/types'
import { createPersist } from '@/utils/persistAdapter'

const persist = createPersist('parformance', 1)

export const useSessionsStore = defineStore('sessions', {
  state: () => ({
    sessions: [] as Session[],
    loaded: false,
  }),
  getters: {
    // Returns all sessions for a given drillId (most recent first)
    listByDrill: (state) => (drillId: string): Session[] =>
      state.sessions
        .filter(s => s.drillId === drillId)
        .sort((a, b) => b.date.localeCompare(a.date)),

    // Returns the latest session for a drillId (by date), or undefined
    latestByDrill(): (drillId: string) => Session | undefined {
      return (drillId: string) => {
        const list = this.listByDrill(drillId)
        return list[0]
      }
    },
  },
  actions: {
    async load() {
      const stored = persist.get<Session[]>('sessions')
      this.sessions = Array.isArray(stored) ? stored : []
      this.loaded = true
    },
    async persist() {
      persist.setDebounced('sessions', this.sessions)
    },
    // Create or insert a session. If id is missing, it will be generated.
    async addSession(session: Omit<Session, 'id'> & Partial<Pick<Session, 'id'>>) {
      const id = session.id ?? crypto.randomUUID()
      const toAdd: Session = { ...session, id }
      this.sessions.push(toAdd)
      // Keep sessions roughly ordered by date desc to make getters cheap (not required)
      this.sessions.sort((a, b) => b.date.localeCompare(a.date))
      await this.persist()
      return toAdd
    },
    async remove(id: string) {
      this.sessions = this.sessions.filter(s => s.id !== id)
      await this.persist()
    },
    async clearAll() {
      this.sessions = []
      await this.persist()
    },
    async updateNotes(id: string, notes: string | undefined | null) {
      const idx = this.sessions.findIndex(s => s.id === id)
      if (idx !== -1) {
        const next = { ...this.sessions[idx], notes: notes ?? undefined }
        this.sessions[idx] = next
        await this.persist()
      }
    }
  }
})
