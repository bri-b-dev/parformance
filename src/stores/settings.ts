import { defineStore } from 'pinia'
import { createPersist } from '@/utils/persistAdapter'

const persist = createPersist('parformance', 1)

export type Handedness = 'right' | 'left'

export interface SettingsState {
  hcp: number | null
  handedness: Handedness
  loaded: boolean
}

export const useSettingsStore = defineStore('settings', {
  state: (): SettingsState => ({
    hcp: null,
    handedness: 'right',
    loaded: false,
  }),
  getters: {
    // direct getters if needed in components
    getHcp: (state) => state.hcp,
    getHandedness: (state) => state.handedness,
  },
  actions: {
    async load() {
      const data = persist.get<Partial<SettingsState>>('settings')
      if (data) {
        if (typeof data.hcp === 'number' || data.hcp === null) this.hcp = data.hcp ?? null
        if (data.handedness === 'right' || data.handedness === 'left') this.handedness = data.handedness
      }
      this.loaded = true
    },
    async persist() {
      persist.setDebounced('settings', { hcp: this.hcp, handedness: this.handedness })
    },
    async setHcp(hcp: number | null) {
      this.hcp = hcp
      // Emit a browser event so components can re-render HCP highlights
      if (typeof window !== 'undefined' && typeof window.dispatchEvent === 'function') {
        let evt: any
        try {
          // Prefer CustomEvent when available (browser)
          // @ts-ignore - CustomEvent may not exist in Node
          evt = new CustomEvent('hcp-changed', { detail: hcp })
        } catch {
          // Fallback for non-browser test environments
          evt = { type: 'hcp-changed', detail: hcp }
        }
        window.dispatchEvent(evt as any)
      }
      await this.persist()
    },
    async setHandedness(value: Handedness) {
      this.handedness = value
      await this.persist()
    },
    async clearAll() {
      this.hcp = null
      this.handedness = 'right'
      await this.persist()
    },
  },
})
