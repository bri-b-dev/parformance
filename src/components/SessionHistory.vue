<template>
  <section class="card" style="margin-top:10px;">
    <header style="display:flex; align-items:center; justify-content:space-between; gap:8px; margin-bottom:6px;">
      <h3 style="margin:0; font-weight:700;">Verlauf</h3>
      <small class="chip" v-if="items.length" aria-label="Anzahl Sessions">{{ items.length }} Einträge</small>
    </header>

    <!-- Empty state -->
    <div v-if="loaded && items.length === 0" class="p-4 text-sm text-gray-600" role="status" aria-live="polite">
      Noch keine Einträge. Starte eine Session und speichere dein Ergebnis.
    </div>

    <!-- List -->
    <ul v-else class="list-none" style="padding:0; margin:0;">
      <li v-for="s in items" :key="s.id" class="card" style="margin:6px 0;">
        <div class="row" style="align-items:center; justify-content:space-between;">
          <div style="display:flex; flex-direction:column; gap:2px;">
            <strong style="line-height:1;">{{ s.result.value }} {{ s.result.unit }}</strong>
            <small style="color:var(--muted);">{{ formatDate(s.date) }}</small>
          </div>
          <div class="chips" style="justify-content:flex-end;">
            <span v-if="s.attempts != null" class="chip" title="Versuche">↻ {{ s.attempts }}</span>
            <span v-if="s.timerUsed != null" class="chip" title="Zeit">⏱️ {{ formatSeconds(s.timerUsed) }}</span>
          </div>
        </div>

        <!-- Notes display / inline edit -->
        <div style="margin-top:6px;">
          <template v-if="editingId === s.id">
            <label class="label" :for="`notes-${s.id}`">Notizen bearbeiten</label>
            <textarea
              :id="`notes-${s.id}`"
              class="input"
              rows="2"
              v-model.trim="draftNotes"
              data-testid="notes-textarea"
              aria-live="polite"
            ></textarea>
            <div class="row" style="justify-content:flex-end; margin-top:6px;">
              <button type="button" class="btn" @click="cancelEdit" data-testid="notes-cancel">Abbrechen</button>
              <button type="button" class="btn btn-primary" @click="saveNotes(s.id)" data-testid="notes-save">Speichern</button>
            </div>
          </template>
          <template v-else>
            <p v-if="s.notes" style="margin:0; white-space:pre-wrap;">{{ s.notes }}</p>
            <p v-else style="margin:0; color:var(--muted);">Keine Notizen</p>
            <div class="row" style="justify-content:flex-end; margin-top:6px;">
              <button type="button" class="btn" @click="startEdit(s)" :aria-label="`Notizen bearbeiten für ${formatDate(s.date)}`" data-testid="notes-edit">Notizen</button>
            </div>
          </template>
        </div>
      </li>
    </ul>
  </section>
</template>

<script setup lang="ts">
import { computed, onMounted, ref } from 'vue'
import type { Session } from '@/types'
import { useSessionsStore } from '@/stores/sessions'

const props = defineProps<{ drillId: string }>()

const sessions = useSessionsStore()

onMounted(async () => {
  if (!sessions.loaded) await sessions.load()
})

const loaded = computed(() => sessions.loaded)
const items = computed(() => sessions.listByDrill(props.drillId))

// Inline notes editing state
const editingId = ref<string | null>(null)
const draftNotes = ref<string>('')

function startEdit(s: Session) {
  editingId.value = s.id
  draftNotes.value = s.notes ?? ''
}
function cancelEdit() {
  editingId.value = null
  draftNotes.value = ''
}
async function saveNotes(id: string) {
  await sessions.updateNotes(id, draftNotes.value || undefined)
  editingId.value = null
  draftNotes.value = ''
}

function pad(n: number) { return n.toString().padStart(2, '0') }
function formatSeconds(total: number) {
  const t = Math.max(0, Math.floor(total || 0))
  const m = Math.floor(t / 60)
  const s = t % 60
  return `${pad(m)}:${pad(s)}`
}
function formatDate(iso: string) {
  // Keep deterministic/simple: YYYY-MM-DD HH:mm (local time)
  try {
    const d = new Date(iso)
    const y = d.getFullYear()
    const m = pad(d.getMonth() + 1)
    const day = pad(d.getDate())
    const hh = pad(d.getHours())
    const mm = pad(d.getMinutes())
    return `${y}-${m}-${day} ${hh}:${mm}`
  } catch {
    return iso
  }
}
</script>
