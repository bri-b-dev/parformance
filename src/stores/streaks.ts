import { defineStore } from 'pinia'
import { useSessionsStore } from './sessions'
import type { Session } from '@/types'
import { computeStreaks } from '@/utils/streaks'

export const useStreaksStore = defineStore('streaks', {
  getters: {
    stats(): { current: number; best: number } {
      const sessionsStore = useSessionsStore()
      const sessions = sessionsStore.sessions as Session[]
      return computeStreaks(
        sessions.map(s => ({ date: s.date }))
      )
    },
    current(): number {
      return this.stats.current
    },
    best(): number {
      return this.stats.best
    }
  },
})

export default { useStreaksStore }
