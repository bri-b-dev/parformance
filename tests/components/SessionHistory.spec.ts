import { describe, it, expect, beforeEach, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import { setActivePinia, createPinia } from 'pinia'
import SessionHistory from '../../src/components/SessionHistory.vue'
import { useSessionsStore } from '../../src/stores/sessions'

class LocalStorageMock {
  store: Record<string, string> = {}
  getItem(key: string) { return this.store[key] ?? null }
  setItem(key: string, value: string) { this.store[key] = value }
  removeItem(key: string) { delete this.store[key] }
  clear() { this.store = {} }
  key(i: number) { return Object.keys(this.store)[i] ?? null }
  get length() { return Object.keys(this.store).length }
}

declare global { var window: any }

describe('SessionHistory', () => {
  beforeEach(() => {
    setActivePinia(createPinia())
    // Patch existing jsdom window instead of replacing it to preserve Event constructors
    Object.assign(global.window, { localStorage: new LocalStorageMock() })
    vi.useFakeTimers()
  })

  it('renders empty state when there are no sessions', async () => {
    const wrapper = mount(SessionHistory, { props: { drillId: 'd1' } })
    const store = useSessionsStore()
    await store.load()

    // Wait next tick for computed to update
    await Promise.resolve()

    expect(wrapper.text()).toContain('Noch keine Einträge')
  })

  it('lists sessions for a drill, sorted by date desc with attempts/timer chips', async () => {
    const store = useSessionsStore()
    await store.load()

    // Seed some sessions across drills
    await store.addSession({ id: 's1', drillId: 'd1', date: '2025-01-01T10:00:00Z', hcp: 12, result: { value: 5, unit: 'pts' }, attempts: 2, notes: 'erste Notiz' })
    await store.addSession({ id: 's2', drillId: 'd1', date: '2025-02-01T10:00:00Z', hcp: 12, result: { value: 7, unit: 'pts' }, timerUsed: 90 })
    await store.addSession({ id: 's3', drillId: 'd2', date: '2025-03-01T10:00:00Z', hcp: 12, result: { value: 9, unit: 'pts' } })

    const wrapper = mount(SessionHistory, { props: { drillId: 'd1' } })

    // Run timers to flush debounced persistence (not strictly required for UI rendering)
    vi.runOnlyPendingTimers()

    const rows = wrapper.findAll('li')
    // Should list only d1 sessions
    expect(rows.length).toBe(2)
    // Sorted desc: s2 first (2025-02...), then s1 (2025-01...)
    expect(wrapper.text()).toMatch(/7 pts[\s\S]*5 pts/)
    // Chips visible
    expect(wrapper.text()).toContain('↻ 2')
    expect(wrapper.text()).toContain('⏱️ 01:30')

    // Snapshot the rendered HTML (Definition of Done)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('supports inline editing of notes and persists via store.updateNotes', async () => {
    const store = useSessionsStore()
    await store.load()

    await store.addSession({ id: 's10', drillId: 'dA', date: '2025-06-01T10:00:00Z', hcp: 18, result: { value: 3, unit: 'pts' } })

    const wrapper = mount(SessionHistory, { props: { drillId: 'dA' } })

    // Click notes edit
    const editBtn = await wrapper.find('[data-testid="notes-edit"]')
    expect(editBtn.exists()).toBe(true)
    await editBtn.trigger('click')

    const ta = wrapper.get('[data-testid="notes-textarea"]')
    await ta.setValue('neue Notiz')

    // Save notes
    const saveBtn = wrapper.get('[data-testid="notes-save"]')
    await saveBtn.trigger('click')
    await wrapper.vm.$nextTick()
    await Promise.resolve()
    await wrapper.vm.$nextTick()

    // Store should be updated
    const saved = store.listByDrill('dA')[0]
    expect(saved.notes).toBe('neue Notiz')

    // UI shows the new notes
    await Promise.resolve()
    await wrapper.vm.$nextTick()
    expect(wrapper.text()).toContain('neue Notiz')
  })
})
