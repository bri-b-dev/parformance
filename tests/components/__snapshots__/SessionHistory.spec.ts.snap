// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SessionHistory > lists sessions for a drill, sorted by date desc with attempts/timer chips 1`] = `
"<section class="card" style="margin-top: 10px;">
  <header style="display: flex; align-items: center; justify-content: space-between; gap: 8px; margin-bottom: 6px;">
    <h3 style="margin: 0px; font-weight: 700;">Verlauf</h3><small class="chip" aria-label="Anzahl Sessions">2 Einträge</small>
  </header><!-- Empty state -->
  <!-- List -->
  <ul class="list-none" style="padding: 0px; margin: 0px;">
    <li class="card" style="margin: 6px 0px;">
      <div class="row" style="align-items: center; justify-content: space-between;">
        <div style="display: flex; flex-direction: column; gap: 2px;"><strong style="line-height: 1;">7 pts</strong><small style="color: var(--muted);">2025-02-01 11:00</small></div>
        <div class="chips" style="justify-content: flex-end;">
          <!--v-if--><span class="chip" title="Zeit">⏱️ 01:30</span>
        </div>
      </div><!-- Notes display / inline edit -->
      <div style="margin-top: 6px;">
        <p style="margin: 0px; color: var(--muted);">Keine Notizen</p>
        <div class="row" style="justify-content: flex-end; margin-top: 6px;"><button type="button" class="btn" aria-label="Notizen bearbeiten für 2025-02-01 11:00" data-testid="notes-edit">Notizen</button></div>
      </div>
    </li>
    <li class="card" style="margin: 6px 0px;">
      <div class="row" style="align-items: center; justify-content: space-between;">
        <div style="display: flex; flex-direction: column; gap: 2px;"><strong style="line-height: 1;">5 pts</strong><small style="color: var(--muted);">2025-01-01 11:00</small></div>
        <div class="chips" style="justify-content: flex-end;"><span class="chip" title="Versuche">↻ 2</span>
          <!--v-if-->
        </div>
      </div><!-- Notes display / inline edit -->
      <div style="margin-top: 6px;">
        <p style="margin: 0px; white-space: pre-wrap;">erste Notiz</p>
        <div class="row" style="justify-content: flex-end; margin-top: 6px;"><button type="button" class="btn" aria-label="Notizen bearbeiten für 2025-01-01 11:00" data-testid="notes-edit">Notizen</button></div>
      </div>
    </li>
  </ul>
</section>"
`;
